storage/innobase/trx/trx0rseg.cc

```diff
@@ -118,6 +134,27 @@ void trx_rseg_update_wsrep_checkpoint(const XID* xid)
 	mlog_write_ull(rseg_header + TRX_RSEG_MAX_TRX_ID,
 		       trx_sys.get_max_trx_id(), &mtr);
 	trx_rseg_update_wsrep_checkpoint(rseg_header, xid, &mtr);
+
+	/* If the UUID part of the wsrep XID does not match to
+	current_xid_uuid, this is either the first call after the
+	startup or wsrep group UUID has changed. Iterate over the
+	rest of the rseg headers and reset wsrep XID to all zeros.
+	*/
+	if (memcmp(current_xid_uuid, read_wsrep_xid_uuid(xid, NULL), 16))
+	{
+		for (ulint rseg_id = 1; rseg_id < TRX_SYS_N_RSEGS; ++rseg_id) {
+			const trx_rseg_t* rseg = trx_sys.rseg_array[rseg_id];
+			trx_rsegf_t* rseg_header = trx_rsegf_get(rseg->space,
+								 rseg->page_no
+								 , &mtr);
+			XID tmp_xid;
+			memset(&tmp_xid, 0, sizeof(tmp_xid));
+			trx_rseg_update_wsrep_checkpoint(rseg_header,
+							 &tmp_xid, &mtr);
+		}
+	}
+	read_wsrep_xid_uuid(xid, current_xid_uuid);
+
 	mtr.commit();
```

Comment:
I don't see any problem in splitting this loop into separate mini-transactions, except that if the server crashes during this operation, the incomplete operation should somehow detected at server initialization. In that case, it should be enough to use first found wsrep XID as a authority and clear the rest of the XIDs whose UUID does not match to the first found. From the implementation it seems that the rseg 0 will always contain the UUID of the last seen group membership change event, so it should be used.

Reply:
Given that the `wsrep_set_SE_checkpoint()` is a rare operation that is executed on DDL operations, we can keep the behaviour that is implemented in the patch. Deadlocks between InnoDB threads should not be an issue, because I think that this is the only operation after startup where multiple rollback segment header pages are concurrently locked.